Sweeper: aws_wafregional_regex_match_set fails when match set is empty

When a `aws_wafregional_regex_match_set ` is empty, the sweeper fails with the following error:

```console
[ERROR] Error running Sweeper (aws_wafregional_regex_match_set) in region (us-west-2): Error updating WAF Regional Regex Match Set: InvalidParameter: 1 validation error(s) found.
 - minimum field size of 1, UpdateRegexMatchSetInput.Updates.
```

### Affected Resource

<!--- Please list the affected resources and data sources. --->

* aws_wafregional_regex_match_set

--- a/aws/internal/service/wafregional/finder/finder.go
+++ b/aws/internal/service/wafregional/finder/finder.go
@@ -1,0 +1,15 @@
+package finder
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/waf"
+	"github.com/aws/aws-sdk-go/service/wafregional"
+)
+
+func RegexMatchSetByID(conn *wafregional.WAFRegional, id string) (*waf.RegexMatchSet, error) {
+	result, err := conn.GetRegexMatchSet(&waf.GetRegexMatchSetInput{
+		RegexMatchSetId: aws.String(id),
+	})
+
+	return result.RegexMatchSet, err
+}

--- a/aws/resource_aws_wafregional_regex_match_set.go
+++ b/aws/resource_aws_wafregional_regex_match_set.go
@@ -9,6 +9,7 @@ import (
 	"github.com/aws/aws-sdk-go/service/waf"
 	"github.com/aws/aws-sdk-go/service/wafregional"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/wafregional/finder"
 )
 
 func resourceAwsWafRegionalRegexMatchSet() *schema.Resource {
@@ -94,12 +95,9 @@ func resourceAwsWafRegionalRegexMatchSetCreate(d *schema.ResourceData, meta inte
 
 func resourceAwsWafRegionalRegexMatchSetRead(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).wafregionalconn
-	log.Printf("[INFO] Reading WAF Regional Regex Match Set: %s", d.Get("name").(string))
-	params := &waf.GetRegexMatchSetInput{
-		RegexMatchSetId: aws.String(d.Id()),
-	}
 
-	resp, err := conn.GetRegexMatchSet(params)
+	log.Printf("[INFO] Reading WAF Regional Regex Match Set: %s", d.Get("name").(string))
+	set, err := finder.RegexMatchSetByID(conn, d.Id())
 	if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
 		log.Printf("[WARN] WAF Regional Regex Match Set (%s) not found, removing from state", d.Id())
 		d.SetId("")
@@ -109,8 +107,8 @@ func resourceAwsWafRegionalRegexMatchSetRead(d *schema.ResourceData, meta interf
 		return fmt.Errorf("Error getting WAF Regional Regex Match Set (%s): %s", d.Id(), err)
 	}
 
-	d.Set("name", resp.RegexMatchSet.Name)
-	d.Set("regex_match_tuple", flattenWafRegexMatchTuples(resp.RegexMatchSet.RegexMatchTuples))
+	d.Set("name", set.Name)
+	d.Set("regex_match_tuple", flattenWafRegexMatchTuples(set.RegexMatchTuples))
 
 	return nil
 }
@@ -142,37 +140,58 @@ func resourceAwsWafRegionalRegexMatchSetDelete(d *schema.ResourceData, meta inte
 	conn := meta.(*AWSClient).wafregionalconn
 	region := meta.(*AWSClient).region
 
-	oldTuples := d.Get("regex_match_tuple").(*schema.Set).List()
-	if len(oldTuples) > 0 {
-		noTuples := []interface{}{}
-		err := updateRegexMatchSetResourceWR(d.Id(), oldTuples, noTuples, conn, region)
-		if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
-			return nil
-		}
-		if err != nil {
-			return fmt.Errorf("Failed updating WAF Regional Regex Match Set(%s): %s", d.Id(), err)
-		}
+	tuples := getRegexMatchTuplesFromResourceData(d)
+	err := clearRegexMatchTuples(conn, region, d.Id(), tuples)
+	if err != nil {
+		return err
 	}
 
-	wr := newWafRegionalRetryer(conn, "global")
-	_, err := wr.RetryWithToken(func(token *string) (interface{}, error) {
-		req := &waf.DeleteRegexMatchSetInput{
-			ChangeToken:     token,
-			RegexMatchSetId: aws.String(d.Id()),
-		}
-		log.Printf("[INFO] Deleting WAF Regional Regex Match Set: %s", req)
-		return conn.DeleteRegexMatchSet(req)
-	})
+	err = deleteRegexMatchSet(conn, "global", d.Id())
 	if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
 		log.Printf("[WARN] WAF Regional Regex Match Set (%s) not found, removing from state", d.Id())
 		d.SetId("")
 		return nil
 	}
 	if err != nil {
-		return fmt.Errorf("Failed deleting WAF Regional Regex Match Set: %s", err)
+		return fmt.Errorf("failed deleting WAF Regional Regex Match Set: %s", err)
 	}
 
 	return nil
+}
+
+func getRegexMatchTuplesFromResourceData(d *schema.ResourceData) []interface{} {
+	return d.Get("regex_match_tuple").(*schema.Set).List()
+}
+
+func getRegexMatchTuplesFromAPIResource(r *waf.RegexMatchSet) []interface{} {
+	return flattenWafRegexMatchTuples(r.RegexMatchTuples)
+}
+
+func clearRegexMatchTuples(conn *wafregional.WAFRegional, region string, id string, oldTuples []interface{}) error {
+	if len(oldTuples) > 0 {
+		noTuples := []interface{}{}
+		err := updateRegexMatchSetResourceWR(id, oldTuples, noTuples, conn, region)
+		if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("failed updating WAF Regional Regex Match Set (%s): %w", id, err)
+		}
+	}
+	return nil
+}
+
+func deleteRegexMatchSet(conn *wafregional.WAFRegional, region, id string) error {
+	log.Printf("[INFO] Deleting WAF Regional Regex Match Set: %s", id)
+	wr := newWafRegionalRetryer(conn, region)
+	_, err := wr.RetryWithToken(func(token *string) (interface{}, error) {
+		req := &waf.DeleteRegexMatchSetInput{
+			ChangeToken:     token,
+			RegexMatchSetId: aws.String(id),
+		}
+		return conn.DeleteRegexMatchSet(req)
+	})
+	return err
 }
 
 func updateRegexMatchSetResourceWR(id string, oldT, newT []interface{}, conn *wafregional.WAFRegional, region string) error {

--- a/aws/resource_aws_wafregional_regex_match_set.go
+++ b/aws/resource_aws_wafregional_regex_match_set.go
@@ -8,6 +8,7 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/waf"
 	"github.com/aws/aws-sdk-go/service/wafregional"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/wafregional/finder"
 )
@@ -98,7 +99,7 @@ func resourceAwsWafRegionalRegexMatchSetRead(d *schema.ResourceData, meta interf
 
 	log.Printf("[INFO] Reading WAF Regional Regex Match Set: %s", d.Get("name").(string))
 	set, err := finder.RegexMatchSetByID(conn, d.Id())
-	if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
+	if tfawserr.ErrCodeEquals(err, wafregional.ErrCodeWAFNonexistentItemException) {
 		log.Printf("[WARN] WAF Regional Regex Match Set (%s) not found, removing from state", d.Id())
 		d.SetId("")
 		return nil
@@ -123,7 +124,7 @@ func resourceAwsWafRegionalRegexMatchSetUpdate(d *schema.ResourceData, meta inte
 		o, n := d.GetChange("regex_match_tuple")
 		oldT, newT := o.(*schema.Set).List(), n.(*schema.Set).List()
 		err := updateRegexMatchSetResourceWR(d.Id(), oldT, newT, conn, region)
-		if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
+		if tfawserr.ErrCodeEquals(err, wafregional.ErrCodeWAFNonexistentItemException) {
 			log.Printf("[WARN] WAF Regional Rate Based Rule (%s) not found, removing from state", d.Id())
 			d.SetId("")
 			return nil
@@ -140,42 +141,53 @@ func resourceAwsWafRegionalRegexMatchSetDelete(d *schema.ResourceData, meta inte
 	conn := meta.(*AWSClient).wafregionalconn
 	region := meta.(*AWSClient).region
 
-	tuples := getRegexMatchTuplesFromResourceData(d)
-	err := clearRegexMatchTuples(conn, region, d.Id(), tuples)
-	if err != nil {
-		return err
-	}
-
-	err = deleteRegexMatchSet(conn, "global", d.Id())
-	if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
+	err := deleteRegexMatchSetResource(conn, region, "global", d.Id(), getRegexMatchTuplesFromResourceData(d))
+	if tfawserr.ErrCodeEquals(err, wafregional.ErrCodeWAFNonexistentItemException) {
 		log.Printf("[WARN] WAF Regional Regex Match Set (%s) not found, removing from state", d.Id())
 		d.SetId("")
 		return nil
 	}
 	if err != nil {
-		return fmt.Errorf("failed deleting WAF Regional Regex Match Set: %s", err)
+		return err
 	}
 
 	return nil
 }
 
-func getRegexMatchTuplesFromResourceData(d *schema.ResourceData) []interface{} {
-	return d.Get("regex_match_tuple").(*schema.Set).List()
+func getRegexMatchTuplesFromResourceData(d *schema.ResourceData) []*waf.RegexMatchTuple {
+	result := []*waf.RegexMatchTuple{}
+
+	for _, t := range d.Get("regex_match_tuple").(*schema.Set).List() {
+		result = append(result, expandWafRegexMatchTuple(t.(map[string]interface{})))
+	}
+
+	return result
 }
 
-func getRegexMatchTuplesFromAPIResource(r *waf.RegexMatchSet) []interface{} {
-	return flattenWafRegexMatchTuples(r.RegexMatchTuples)
+func getRegexMatchTuplesFromAPIResource(r *waf.RegexMatchSet) []*waf.RegexMatchTuple {
+	return r.RegexMatchTuples
 }
 
-func clearRegexMatchTuples(conn *wafregional.WAFRegional, region string, id string, oldTuples []interface{}) error {
-	if len(oldTuples) > 0 {
-		noTuples := []interface{}{}
-		err := updateRegexMatchSetResourceWR(id, oldTuples, noTuples, conn, region)
-		if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
-			return nil
+func clearRegexMatchTuples(conn *wafregional.WAFRegional, region string, id string, tuples []*waf.RegexMatchTuple) error {
+	if len(tuples) > 0 {
+		input := &waf.UpdateRegexMatchSetInput{
+			RegexMatchSetId: aws.String(id),
 		}
+		for _, tuple := range tuples {
+			input.Updates = append(input.Updates, &waf.RegexMatchSetUpdate{
+				Action:          aws.String(waf.ChangeActionDelete),
+				RegexMatchTuple: tuple,
+			})
+		}
+
+		log.Printf("[INFO] Clearing WAF Regional Regex Match Set: %s", id)
+		wr := newWafRegionalRetryer(conn, region)
+		_, err := wr.RetryWithToken(func(token *string) (interface{}, error) {
+			input.ChangeToken = token
+			return conn.UpdateRegexMatchSet(input)
+		})
 		if err != nil {
-			return fmt.Errorf("failed updating WAF Regional Regex Match Set (%s): %w", id, err)
+			return fmt.Errorf("error clearing WAF Regional Regex Match Set (%s): %w", id, err)
 		}
 	}
 	return nil
@@ -191,7 +203,19 @@ func deleteRegexMatchSet(conn *wafregional.WAFRegional, region, id string) error
 		}
 		return conn.DeleteRegexMatchSet(req)
 	})
-	return err
+	if err != nil {
+		return fmt.Errorf("error deleting WAF Regional Regex Match Set (%s): %w", id, err)
+	}
+	return nil
+}
+
+func deleteRegexMatchSetResource(conn *wafregional.WAFRegional, region, tokenRegion, id string, tuples []*waf.RegexMatchTuple) error {
+	err := clearRegexMatchTuples(conn, region, id, tuples)
+	if err != nil {
+		return err
+	}
+
+	return deleteRegexMatchSet(conn, tokenRegion, id)
 }
 
 func updateRegexMatchSetResourceWR(id string, oldT, newT []interface{}, conn *wafregional.WAFRegional, region string) error {

--- a/aws/resource_aws_wafregional_regex_match_set_test.go
+++ b/aws/resource_aws_wafregional_regex_match_set_test.go
@@ -8,9 +8,11 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/waf"
 	"github.com/aws/aws-sdk-go/service/wafregional"
+	"github.com/hashicorp/go-multierror"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/wafregional/finder"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfawsresource"
 )
 
@@ -28,51 +30,64 @@ func testSweepWafRegionalRegexMatchSet(region string) error {
 	}
 	conn := client.(*AWSClient).wafregionalconn
 
-	req := &waf.ListRegexMatchSetsInput{}
-	resp, err := conn.ListRegexMatchSets(req)
-	if err != nil {
-		if testSweepSkipSweepError(err) {
-			log.Printf("[WARN] Skipping WAF Regional Regex Match Set sweep for %s: %s", region, err)
-			return nil
+	var sweeperErrs *multierror.Error
+
+	err = ListRegexMatchSetsPages(conn, &waf.ListRegexMatchSetsInput{}, func(page *waf.ListRegexMatchSetsOutput, lastPage bool) bool {
+		if page == nil {
+			return !lastPage
 		}
-		return fmt.Errorf("Error describing WAF Regional Regex Match Sets: %s", err)
+
+		for _, r := range page.RegexMatchSets {
+			id := aws.StringValue(r.RegexMatchSetId)
+
+			set, err := finder.RegexMatchSetByID(conn, id)
+			if err != nil {
+				sweeperErrs = multierror.Append(sweeperErrs, fmt.Errorf("error retrieving WAF Regional Regex Match Set (%s): %w", id, err))
+				continue
+			}
+
+			tuples := getRegexMatchTuplesFromAPIResource(set)
+			err = clearRegexMatchTuples(conn, region, id, tuples)
+			if err != nil {
+				sweeperErrs = multierror.Append(sweeperErrs, fmt.Errorf("error clearing WAF Regional Regex Match Set (%s): %w", id, err))
+				continue
+			}
+
+			err = deleteRegexMatchSet(conn, region, id)
+			if err != nil {
+				sweeperErrs = multierror.Append(sweeperErrs, fmt.Errorf("error deleting WAF Regional Regex Match Set (%s): %w", id, err))
+				continue
+			}
+		}
+
+		return !lastPage
+	})
+
+	if testSweepSkipSweepError(err) {
+		log.Printf("[WARN] Skipping WAF Regional Regex Match Set sweep for %s: %s", region, err)
+		return sweeperErrs.ErrorOrNil() // In case we have completed some pages, but had errors
+	}
+	if err != nil {
+		sweeperErrs = multierror.Append(sweeperErrs, fmt.Errorf("error describing WAF Regional Regex Match Sets: %w", err))
 	}
 
-	if len(resp.RegexMatchSets) == 0 {
-		log.Print("[DEBUG] No AWS WAF Regional Regex Match Sets to sweep")
-		return nil
-	}
+	return sweeperErrs.ErrorOrNil()
+}
 
-	for _, s := range resp.RegexMatchSets {
-		resp, err := conn.GetRegexMatchSet(&waf.GetRegexMatchSetInput{
-			RegexMatchSetId: s.RegexMatchSetId,
-		})
+func ListRegexMatchSetsPages(conn *wafregional.WAFRegional, input *waf.ListRegexMatchSetsInput, fn func(*waf.ListRegexMatchSetsOutput, bool) bool) error {
+	for {
+		output, err := conn.ListRegexMatchSets(input)
 		if err != nil {
 			return err
 		}
-		set := resp.RegexMatchSet
 
-		oldTuples := flattenWafRegexMatchTuples(set.RegexMatchTuples)
-		noTuples := []interface{}{}
-		err = updateRegexMatchSetResourceWR(*set.RegexMatchSetId, oldTuples, noTuples, conn, region)
-		if err != nil {
-			return fmt.Errorf("Error updating WAF Regional Regex Match Set: %s", err)
+		lastPage := aws.StringValue(output.NextMarker) == ""
+		if !fn(output, lastPage) || lastPage {
+			break
 		}
 
-		wr := newWafRegionalRetryer(conn, region)
-		_, err = wr.RetryWithToken(func(token *string) (interface{}, error) {
-			req := &waf.DeleteRegexMatchSetInput{
-				ChangeToken:     token,
-				RegexMatchSetId: set.RegexMatchSetId,
-			}
-			log.Printf("[INFO] Deleting WAF Regional Regex Match Set: %s", req)
-			return conn.DeleteRegexMatchSet(req)
-		})
-		if err != nil {
-			return fmt.Errorf("error deleting WAF Regional Regex Match Set (%s): %s", aws.StringValue(set.RegexMatchSetId), err)
-		}
+		input.NextMarker = output.NextMarker
 	}
-
 	return nil
 }
 

--- a/aws/resource_aws_wafregional_regex_match_set_test.go
+++ b/aws/resource_aws_wafregional_regex_match_set_test.go
@@ -8,6 +8,7 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/waf"
 	"github.com/aws/aws-sdk-go/service/wafregional"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/go-multierror"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
@@ -46,16 +47,11 @@ func testSweepWafRegionalRegexMatchSet(region string) error {
 				continue
 			}
 
-			tuples := getRegexMatchTuplesFromAPIResource(set)
-			err = clearRegexMatchTuples(conn, region, id, tuples)
+			err = deleteRegexMatchSetResource(conn, region, region, id, getRegexMatchTuplesFromAPIResource(set))
 			if err != nil {
-				sweeperErrs = multierror.Append(sweeperErrs, fmt.Errorf("error clearing WAF Regional Regex Match Set (%s): %w", id, err))
-				continue
-			}
-
-			err = deleteRegexMatchSet(conn, region, id)
-			if err != nil {
-				sweeperErrs = multierror.Append(sweeperErrs, fmt.Errorf("error deleting WAF Regional Regex Match Set (%s): %w", id, err))
+				if !tfawserr.ErrCodeEquals(err, wafregional.ErrCodeWAFNonexistentItemException) {
+					sweeperErrs = multierror.Append(sweeperErrs, fmt.Errorf("error deleting WAF Regional Regex Match Set (%s): %w", id, err))
+				}
 				continue
 			}
 		}
@@ -250,52 +246,12 @@ func testAccAWSWafRegionalRegexMatchSet_disappears(t *testing.T) {
 				Config: testAccAWSWafRegionalRegexMatchSetConfig(matchSetName, patternSetName),
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckAWSWafRegionalRegexMatchSetExists(resourceName, &matchSet),
-					testAccCheckAWSWafRegionalRegexMatchSetDisappears(&matchSet),
+					testAccCheckResourceDisappears(testAccProvider, resourceAwsWafRegionalRegexMatchSet(), resourceName),
 				),
 				ExpectNonEmptyPlan: true,
 			},
 		},
 	})
-}
-
-func testAccCheckAWSWafRegionalRegexMatchSetDisappears(set *waf.RegexMatchSet) resource.TestCheckFunc {
-	return func(s *terraform.State) error {
-		conn := testAccProvider.Meta().(*AWSClient).wafregionalconn
-		region := testAccProvider.Meta().(*AWSClient).region
-
-		wr := newWafRegionalRetryer(conn, region)
-		_, err := wr.RetryWithToken(func(token *string) (interface{}, error) {
-			req := &waf.UpdateRegexMatchSetInput{
-				ChangeToken:     token,
-				RegexMatchSetId: set.RegexMatchSetId,
-			}
-
-			for _, tuple := range set.RegexMatchTuples {
-				req.Updates = append(req.Updates, &waf.RegexMatchSetUpdate{
-					Action:          aws.String("DELETE"),
-					RegexMatchTuple: tuple,
-				})
-			}
-
-			return conn.UpdateRegexMatchSet(req)
-		})
-		if err != nil {
-			return fmt.Errorf("Failed updating WAF Regional Regex Match Set: %s", err)
-		}
-
-		_, err = wr.RetryWithToken(func(token *string) (interface{}, error) {
-			opts := &waf.DeleteRegexMatchSetInput{
-				ChangeToken:     token,
-				RegexMatchSetId: set.RegexMatchSetId,
-			}
-			return conn.DeleteRegexMatchSet(opts)
-		})
-		if err != nil {
-			return fmt.Errorf("Failed deleting WAF Regional Regex Match Set: %s", err)
-		}
-
-		return nil
-	}
 }
 
 func testAccCheckAWSWafRegionalRegexMatchSetExists(n string, v *waf.RegexMatchSet) resource.TestCheckFunc {
@@ -345,7 +301,7 @@ func testAccCheckAWSWafRegionalRegexMatchSetDestroy(s *terraform.State) error {
 		}
 
 		// Return nil if the Regex Pattern Set is already destroyed
-		if isAWSErr(err, wafregional.ErrCodeWAFNonexistentItemException, "") {
+		if tfawserr.ErrCodeEquals(err, wafregional.ErrCodeWAFNonexistentItemException) {
 			return nil
 		}
 

--- a/aws/waf_helpers.go
+++ b/aws/waf_helpers.go
@@ -310,6 +310,15 @@ func flattenWafRegexMatchTuples(tuples []*waf.RegexMatchTuple) []interface{} {
 	return out
 }
 
+func expandWafRegexMatchTuple(tuple map[string]interface{}) *waf.RegexMatchTuple {
+	ftm := tuple["field_to_match"].([]interface{})
+	return &waf.RegexMatchTuple{
+		FieldToMatch:       expandFieldToMatch(ftm[0].(map[string]interface{})),
+		RegexPatternSetId:  aws.String(tuple["regex_pattern_set_id"].(string)),
+		TextTransformation: aws.String(tuple["text_transformation"].(string)),
+	}
+}
+
 func diffWafRegexMatchSetTuples(oldT, newT []interface{}) []*waf.RegexMatchSetUpdate {
 	updates := make([]*waf.RegexMatchSetUpdate, 0)
 
@@ -321,28 +330,18 @@ func diffWafRegexMatchSetTuples(oldT, newT []interface{}) []*waf.RegexMatchSetUp
 			continue
 		}
 
-		ftm := tuple["field_to_match"].([]interface{})
 		updates = append(updates, &waf.RegexMatchSetUpdate{
-			Action: aws.String(waf.ChangeActionDelete),
-			RegexMatchTuple: &waf.RegexMatchTuple{
-				FieldToMatch:       expandFieldToMatch(ftm[0].(map[string]interface{})),
-				RegexPatternSetId:  aws.String(tuple["regex_pattern_set_id"].(string)),
-				TextTransformation: aws.String(tuple["text_transformation"].(string)),
-			},
+			Action:          aws.String(waf.ChangeActionDelete),
+			RegexMatchTuple: expandWafRegexMatchTuple(tuple),
 		})
 	}
 
 	for _, nt := range newT {
 		tuple := nt.(map[string]interface{})
 
-		ftm := tuple["field_to_match"].([]interface{})
 		updates = append(updates, &waf.RegexMatchSetUpdate{
-			Action: aws.String(waf.ChangeActionInsert),
-			RegexMatchTuple: &waf.RegexMatchTuple{
-				FieldToMatch:       expandFieldToMatch(ftm[0].(map[string]interface{})),
-				RegexPatternSetId:  aws.String(tuple["regex_pattern_set_id"].(string)),
-				TextTransformation: aws.String(tuple["text_transformation"].(string)),
-			},
+			Action:          aws.String(waf.ChangeActionInsert),
+			RegexMatchTuple: expandWafRegexMatchTuple(tuple),
 		})
 	}
 	return updates

--- a/aws/waf_token_handlers.go
+++ b/aws/waf_token_handlers.go
@@ -4,8 +4,8 @@ import (
 	"fmt"
 	"time"
 
-	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/waf"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 )
 
@@ -25,13 +25,12 @@ func (t *WafRetryer) RetryWithToken(f withTokenFunc) (interface{}, error) {
 		var err error
 		tokenOut, err = t.Connection.GetChangeToken(&waf.GetChangeTokenInput{})
 		if err != nil {
-			return resource.NonRetryableError(fmt.Errorf("Failed to acquire change token: %s", err))
+			return resource.NonRetryableError(fmt.Errorf("Failed to acquire change token: %w", err))
 		}
 
 		out, err = f(tokenOut.ChangeToken)
 		if err != nil {
-			awsErr, ok := err.(awserr.Error)
-			if ok && awsErr.Code() == "WAFStaleDataException" {
+			if tfawserr.ErrCodeEquals(err, waf.ErrCodeStaleDataException) {
 				return resource.RetryableError(err)
 			}
 			return resource.NonRetryableError(err)
@@ -42,7 +41,7 @@ func (t *WafRetryer) RetryWithToken(f withTokenFunc) (interface{}, error) {
 		tokenOut, err = t.Connection.GetChangeToken(&waf.GetChangeTokenInput{})
 
 		if err != nil {
-			return nil, fmt.Errorf("error getting WAF change token: %s", err)
+			return nil, fmt.Errorf("error getting WAF change token: %w", err)
 		}
 
 		out, err = f(tokenOut.ChangeToken)

--- a/aws/wafregional_token_handlers.go
+++ b/aws/wafregional_token_handlers.go
@@ -4,9 +4,9 @@ import (
 	"fmt"
 	"time"
 
-	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/waf"
 	"github.com/aws/aws-sdk-go/service/wafregional"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 )
 
@@ -28,13 +28,12 @@ func (t *WafRegionalRetryer) RetryWithToken(f withRegionalTokenFunc) (interface{
 
 		tokenOut, err = t.Connection.GetChangeToken(&waf.GetChangeTokenInput{})
 		if err != nil {
-			return resource.NonRetryableError(fmt.Errorf("Failed to acquire change token: %s", err))
+			return resource.NonRetryableError(fmt.Errorf("Failed to acquire change token: %w", err))
 		}
 
 		out, err = f(tokenOut.ChangeToken)
 		if err != nil {
-			awsErr, ok := err.(awserr.Error)
-			if ok && awsErr.Code() == "WAFStaleDataException" {
+			if tfawserr.ErrCodeEquals(err, waf.ErrCodeStaleDataException) {
 				return resource.RetryableError(err)
 			}
 			return resource.NonRetryableError(err)
@@ -45,7 +44,7 @@ func (t *WafRegionalRetryer) RetryWithToken(f withRegionalTokenFunc) (interface{
 		tokenOut, err = t.Connection.GetChangeToken(&waf.GetChangeTokenInput{})
 
 		if err != nil {
-			return nil, fmt.Errorf("error getting WAF Regional change token: %s", err)
+			return nil, fmt.Errorf("error getting WAF Regional change token: %w", err)
 		}
 
 		out, err = f(tokenOut.ChangeToken)

